main/src/library/File.flix

```diff
@@ -840,4 +840,104 @@ namespace File {
+                Err(getMessage(ex))
+        }
+
+    ///
+    /// Returns an Iterator naming the files and directories in the directory `f`.
+    ///
+    /// Does not recursively traverse the directory.
+    ///
+    pub def listWithIter(f: String): Result[Iterator[String], String] & Impure =
+        try {
+            import new java.io.File(String): ##java.io.File & Impure as newFile;
+            import java.io.File.toPath(): ##java.nio.file.Path & Impure;
+            import static java.nio.file.Files.newDirectoryStream(##java.nio.file.Path): ##java.nio.file.DirectoryStream & Impure;
+            import java.nio.file.DirectoryStream.iterator(): ##java.util.Iterator & Impure;
+            import java.nio.file.Path.toString(): String & Impure as pathToString;
+
+            let javaFile = newFile(f);
+            let javaPath = toPath(javaFile);
+            let javaStream = newDirectoryStream(javaPath);
+            let javaIter = iterator(javaStream);
+
+            use Result.flatMap;
+            let* Iterator(done, next) = fromJavaIter(javaIter);
+            let next1 = () -> pathToString(next());
+            Ok(Iterator(done, next1))
+        } catch {
+            case ex: ##java.io.IOException =>
+                import java.lang.Throwable.getMessage(): String & Impure;
+                Err(getMessage(ex))
+        }
+
+    ///
+    /// Returns an Iterator naming the files and directories in the directory `f`.
+    ///
+    /// Recursively traverses the directory.
+    ///
+    pub def walk(f: String): Result[Iterator[String], String] & Impure =
+        try {
+            import new java.io.File(String): ##java.io.File & Impure as newFile;
+            import java.io.File.toPath(): ##java.nio.file.Path & Impure;
+            import static java.nio.file.Files.walk(##java.nio.file.Path, Array[##java.nio.file.FileVisitOption]): ##java.util.stream.Stream & Impure;
+            import java.util.stream.BaseStream.iterator(): ##java.util.Iterator & Impure;
+            import java.nio.file.Path.getFileName(): ##java.nio.file.Path & Impure;
+            import java.nio.file.Path.toString(): String & Impure as pathToString;
+
+            let javaFile = newFile(f);
+            let javaPath = toPath(javaFile);
+            let javaStream = walk(javaPath, []);
+            let javaIter = iterator(javaStream as ##java.util.stream.BaseStream);
+
+            use Result.flatMap;
+            let* Iterator(done, next) = fromJavaIter(javaIter);
+            let next1 = () -> pathToString(getFileName(next()));
+            Ok(Iterator(done, next1))
+        } catch {
+            case ex: ##java.io.IOException =>
+                import java.lang.Throwable.getMessage(): String & Impure;
+                Err(getMessage(ex))
+        }
+
+    ///
+    /// Converts a java iterator to an Iterator.
+    ///
+    pub def fromJavaIter(iter: ##java.util.Iterator): Result[Iterator[a], String] & Impure =
```

Comment:
How about this?

Reply:
This looks good, but I don't think it needs to return a result, right?